fireplace/cards/utils.py

```diff
@@ -38,6 +38,50 @@
 AT_MAX_MANA = lambda s: MANA(s) == 10
 
 
+# take a list of weights and a list of card pools and produce a random weighted sample without replacement
+def weighted_card_choice(source, weights, card_sets, count):
+		# get all the cards matching each set of filters and create the weight line
+		# (user does not have to normalize weighting from 0-1)
+		cardlist = []
+		weightedsegmentstarts = [0] * len(card_sets)
+		weightlinelength = 0
+
+		for i, c in enumerate(card_sets):
+			weightedsegmentstarts[i] = len(cardlist)
+			cardlist += c
+			# higher weighted sets (segments) occupy more space per item on the weight line
+			weightlinelength += len(c) * weights[i]
+
+		ret = []
+
+		# for each card
+		for c in range(count):
+
+			# pick a point along the weight line
+			r = random.random() * weightlinelength
```

Comment:
That would give a small chance of r == weightlinelength which would be out of bounds. random.random() will never return 1.0.